### PR TITLE
Mr13 implement mining

### PR DESCRIPTION
**Pull request by baldhat**

### CODE REVIEW

File: src/main/kotlin/com/msd/robot/application/RobotApplicationService.kt [hunk 1]
```diff
@@ -214,4 +220,213 @@ class RobotApplicationService(
...
+     * 2. the planet on which the robot is positioned exists and has a resource patch on it
+     * 3. the robot has the necessary mining level to mine the resource on the planet
+     *
+     * @param mineCommands: The list of [MineCommand]s which gets filtered for validity
+     * @param planetsToResources: A map of planets assigning each a resourceType or a null value, representing no
+     *                            resource present on the planet.
+     * @return a list of [ValidMineCommand]s, representing only the MineCommands which are valid and having replaced
+*              the IDs with the corresponding entities.
+     */
+    private fun replaceIdsByObjectsInValidMineCommands(
```
Comment (DennisWae) [inline src/main/kotlin/com/msd/robot/application/RobotApplicationService.kt:284]: Should be named `replacedIdsWithObjectsInValidMineCommands`

File: src/main/kotlin/com/msd/robot/application/RobotApplicationService.kt [hunk 2]
```diff
@@ -214,4 +220,213 @@ class RobotApplicationService(
...
+        planetsToResources: Map<UUID, ResourceType?>
+    ): MutableList<ValidMineCommand> {
+        val validMineCommands = mutableListOf<ValidMineCommand>()
+
+        for (mineCommand in mineCommands) {
+            try {
+                val robot = robotDomainService.getRobot(mineCommand.robotUUID)
+                val resource = planetsToResources[robot.planet.planetId]
+                    ?: throw NoResourceOnPlanetException(robot.planet.planetId)
+                val extendedMineCommand = ValidMineCommand(
```
Comment (DennisWae) [inline src/main/kotlin/com/msd/robot/application/RobotApplicationService.kt:295]: Why is this variable named extendedMineCommand

**Review by DennisWae — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/msd/application/MineRequestDto.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.msd.application
+
+class MineRequestDto(
+    amount: Int
+) {
+    val mining: MiningDto = MiningDto(amount)
+
+    class MiningDto(
+        amount_requested: Int
```
[the remark this answers is not in the document]
Comment (baldhat) [reply src/main/kotlin/com/msd/application/MineRequestDto.kt:9]: This is just the expected structure of the json by the Map API, it expects a json object with the name mining, containing the amount_requested

File: src/main/kotlin/com/msd/robot/application/InvalidPlanetException.kt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package com.msd.robot.application
+
+import org.springframework.http.HttpStatus
+import org.springframework.web.bind.annotation.ResponseStatus
+import java.util.*
+
+@ResponseStatus(HttpStatus.BAD_REQUEST)
+class InvalidPlanetException(planet: UUID) : RuntimeException("Invalid planet with UUID $planet")
```
[the remark this answers is not in the document]
Comment (baldhat) [reply src/main/kotlin/com/msd/robot/application/InvalidPlanetException.kt]: no longer used, removed it

File: src/main/kotlin/com/msd/robot/application/RobotApplicationService.kt [hunk 5]
```diff
@@ -214,4 +220,213 @@ class RobotApplicationService(
...
+    /**
+     * Create a map of planets and the resource each offers. Each planet either has a single resource or none,
+     * represented with a null value.
+     *
+     * @param mineCommands: A list of commands, for which the resources should be fetched.
+     * @return a map assigning each planet a resource or a null value
+     */
+    private fun getResourcesOnPlanets(mineCommands: List<MineCommand>) = mineCommands
+        .map {
+            try {
```
[the remark this answers is not in the document]
Comment (baldhat) [reply src/main/kotlin/com/msd/robot/application/RobotApplicationService.kt:250]: In this case it should not.

File: src/main/kotlin/com/msd/robot/application/RobotApplicationService.kt [hunk 6]
```diff
@@ -214,4 +220,213 @@ class RobotApplicationService(
...
+    }
+
+    /**
+     * Create a map of planets and the resource each offers. Each planet either has a single resource or none,
+     * represented with a null value.
+     *
+     * @param mineCommands: A list of commands, for which the resources should be fetched.
+     * @return a map assigning each planet a resource or a null value
+     */
+    private fun getResourcesOnPlanets(mineCommands: List<MineCommand>) = mineCommands
```
[the remark this answers is not in the document]
Comment (baldhat) [reply src/main/kotlin/com/msd/robot/application/RobotApplicationService.kt]: done

File: src/test/kotlin/com/msd/robot/domain/InventoryTest.kt [hunk 7]
```diff
@@ -215,4 +215,14 @@ class InventoryTest {
             }
         )
     }
+
+//    @Test
+//    fun `Mining correctly adds resources to inventory`() {
+//        // given
+//        assertEquals(0, robot1.inventory.getStorageUsageForResource(ResourceType.IRON))
```
[the remark this answers is not in the document]
Comment (baldhat) [reply src/test/kotlin/com/msd/robot/domain/InventoryTest.kt]: This used the old robot.mine( ) method, I'll delete the code/comment

File: src/main/kotlin/com/msd/application/GameMapPlanetDto.kt [hunk 8]
```diff
@@ -5,7 +5,8 @@ import java.util.*
 
 class GameMapPlanetDto(
     val id: UUID,
-    val movementCost: Int
+    val movement_difficulty: Int,
```
[the remark this answers is not in the document]
Comment (baldhat) [reply src/main/kotlin/com/msd/application/GameMapPlanetDto.kt]: changed the name to movementDifficulty, and added the jsonproperty

File: src/main/kotlin/com/msd/application/ResourceDto.kt [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package com.msd.application
+
+import com.msd.domain.ResourceType
+
+class ResourceDto(
+    val resource_type: ResourceType
```
[the remark this answers is not in the document]
Comment (baldhat) [reply src/main/kotlin/com/msd/application/ResourceDto.kt]: done

File: src/main/kotlin/com/msd/robot/application/RobotApplicationService.kt (hunk 2 above)
Comment (baldhat) [reply src/main/kotlin/com/msd/robot/application/RobotApplicationService.kt]: changed it

**Pull request merged**